modules/engine/src/lib/shader-factory.ts

```diff
@@ -0,0 +1,52 @@
+import {Device, Shader, ShaderProps} from '@luma.gl/core';
+
+/** Manages a cached pool of Shaders for reuse. */
+export class ShaderFactory {
+  static readonly defaultProps: Required<ShaderProps> = {...Shader.defaultProps};
+
+  public readonly device: Device;
+
+  private readonly _shaderCache: Record<string, Shader> = {};
```

Comment:
This part always gave me pause in the PipelineFactory... I wonder if a single record per key isn't more readable? (I hope you are not just following my bad examples :)

```suggestion
  private readonly _cache: Record<string, {shader: Shader, useCount: number}> = {};
```